syncapi/sync/requestpool.go

```diff
@@ -37,18 +40,63 @@ import (
 // RequestPool manages HTTP long-poll connections for /sync
 type RequestPool struct {
 	db       storage.Database
+	cfg      *config.SyncAPI
 	userAPI  userapi.UserInternalAPI
 	notifier *Notifier
 	keyAPI   keyapi.KeyInternalAPI
 	rsAPI    roomserverAPI.RoomserverInternalAPI
+	lastseen sync.Map
 }
 
 // NewRequestPool makes a new RequestPool
 func NewRequestPool(
-	db storage.Database, n *Notifier, userAPI userapi.UserInternalAPI, keyAPI keyapi.KeyInternalAPI,
+	db storage.Database, cfg *config.SyncAPI, n *Notifier,
+	userAPI userapi.UserInternalAPI, keyAPI keyapi.KeyInternalAPI,
 	rsAPI roomserverAPI.RoomserverInternalAPI,
 ) *RequestPool {
-	return &RequestPool{db, userAPI, n, keyAPI, rsAPI}
+	rp := &RequestPool{db, cfg, userAPI, n, keyAPI, rsAPI, sync.Map{}}
+	go rp.cleanLastSeen()
+	return rp
+}
+
+func (rp *RequestPool) cleanLastSeen() {
+	for {
+		rp.lastseen.Range(func(key interface{}, value interface{}) bool {
+			lastseen := value.(time.Time)
+			if time.Since(lastseen) > time.Minute*2 {
+				rp.lastseen.Delete(key)
```

Comment:
I was expecting a slightly simpler method I guess:
 - When a new request comes in, just blindly store (key=`device.UserID+device.ID`, value=`req.RemoteAddr`) all the time.
 - This cleaner goroutine ranges every minute and deletes _everything_, converting each item into a goroutine request to the user API.

This does mean that some updates will be done quicker than others, but no update should take longer than the refresh interval (1min).

Reply:
I can certainly do that.